File: prototype/rails/src/test/models/fake_card_test.rb
require "test_helper"

class FakeCardTest < ActiveSupport::TestCase
  test "the approved number is approved" do
    assert_predicate FakeCard.new("4242424242424242"), :approved?
  end

  test "spaces and dashes are ignored" do
    assert_predicate FakeCard.new("4242 4242-4242 4242"), :approved?
  end

  test "an approved card carries no decline reason" do
    assert_nil FakeCard.new("4242424242424242").decline_reason
  end

  test "the generic decline number is declined" do
    card = FakeCard.new("4000 0000 0000 0002")

    refute_predicate card, :approved?
    assert_equal "generic_decline", card.decline_reason
  end

  test "the insufficient funds number is declined" do
    assert_equal "insufficient_funds", FakeCard.new("4000 0000 0000 9995").decline_reason
  end

  test "any other number is not a valid card" do
    assert_equal "invalid_card_number", FakeCard.new("1234 5678 1234 5678").decline_reason
  end

  test "only the last four digits come back" do
    assert_equal "9995", FakeCard.new("4000 0000 0000 9995").last_four
  end

  test "a number shorter than four digits keeps what it has" do
    assert_equal "12", FakeCard.new("12").last_four
  end
end

File: prototype/node/src/app/core/payments/fake-card.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { decideCard } from './fake-card.ts'
import { isCardApproved } from './card-decision.ts'

test('the approved number is approved', () => {
  assert.equal(isCardApproved(decideCard('4242424242424242')), true)
})

test('spaces and dashes are ignored', () => {
  assert.equal(isCardApproved(decideCard('4242 4242-4242 4242')), true)
})

test('the generic decline number is declined', () => {
  const decision = decideCard('4000 0000 0000 0002')

  assert.equal(isCardApproved(decision), false)
  assert.equal(decision.declineReason, 'generic_decline')
})

test('the insufficient funds number is declined', () => {
  assert.equal(decideCard('4000 0000 0000 9995').declineReason, 'insufficient_funds')
})

test('any other number is not a valid card', () => {
  assert.equal(decideCard('1234 5678 1234 5678').declineReason, 'invalid_card_number')
})

test('only the last four digits come back', () => {
  assert.equal(decideCard('4000 0000 0000 9995').lastFour, '9995')
})

test('a number shorter than four digits keeps what it has', () => {
  assert.equal(decideCard('12').lastFour, '12')
})
